BJS2-27008

amilosh opened the pull request:
- add CRUD endpoints to operate Goals
- add 'find goals' end-points with filters

amilosh-softeq committed:
BJS2-27008
- add CRUD endpoints to operate Goals
- add 'find goals' end-points with filters

IvanTsegelnyuk commented on src/main/java/school/faang/user_service/service/goal/GoalValidator.java:
```diff
@@ -0,0 +1,67 @@
+    private final GoalRepository goalRepository;
+    private final SkillRepository skillRepository;
+
+    private void validate(GoalDto goalDto) {
```
private методы всегда же ниже публичных? ну либо приватный сразу ниже за публичным, в котором он используется. Или нет?

amilosh replied:
Исправил.

amilosh-softeq committed:
BJS2-27008
- use Entity in services instead of EntityDto

kripersh27 commented on src/main/java/school/faang/user_service/dto/goal/GoalDto.java:
```diff
@@ -0,0 +1,47 @@
+package school.faang.user_service.dto.goal;
+
+import com.fasterxml.jackson.annotation.JsonSetter;
+import com.fasterxml.jackson.annotation.Nulls;
```
Серые иморты остались

amilosh replied:
Исправил.

kripersh27 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+        goalService.deleteGoal(goalId);
+    }
+
+    @GetMapping("/sub-goals/{parentGoalId}")
+    public List<GoalDto> findSubGoalsByParentGoalId(@PathVariable Long parentGoalId, @RequestBody GoalFilterDto filter) {
+        List<Goal> foundGoals = goalService.findSubGoalsByParentGoalId(parentGoalId, filter);
+        return goalMapper.toDtos(foundGoals);
+    }
```
Гет по спецификации не имеет тела, может Post тогда сделать?

amilosh replied:
Убрал @RequestBody. Спринг умеет скукоживать параметры из url в объект.

kripersh27 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+        return goalMapper.toDtos(foundGoals);
+    }
+
+    @GetMapping("/user/{userId}")
+    public List<GoalDto> findGoalsByUser(@PathVariable long userId, @RequestBody GoalFilterDto filter) {
+        List<Goal> foundGoals = goalService.findGoalsByUserId(userId, filter);
+        return goalMapper.toDtos(foundGoals);
+    }
```
Тоже самое про Post

kripersh27 commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+        List<Skill> skills = skillRepository.findSkillsByGoalId(goal.getId());
+        Goal foundGoal = goalRepository.findById(goal.getId()).get();
+        foundGoal.getUsers().forEach(user -> {
+            for (Skill skill : skills) {
+                skillRepository.assignSkillToUser(skill.getId(), user.getId());
+            }
```
foreach вместо цикла можно

amilosh replied:
Исправил.

kripersh27 commented on src/main/java/school/faang/user_service/dto/goal/GoalDto.java:
```diff
@@ -0,0 +1,47 @@
+
+    @NotNull(message = "Description cannot be null", groups = {Create.class, Update.class})
+    @NotBlank(message = "Description cannot be blank", groups = {Create.class, Update.class})
+    @Size(max = 128, message = "Description cannot be longer than 64 characters", groups = {Create.class, Update.class})
```
Максимальный размер 128, а в сообщении указано только 64

amilosh replied:
Исправил.

kripersh27 reviewed (commented):
Что нашел - указал)

amilosh-softeq committed:
BJS2-27008: make some updates after peer review
- for GET request use pojo instead of @RequestBody
- remove userId from 'update Goal' functionality

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalValidator.java:
```diff
@@ -0,0 +1,74 @@
+import static school.faang.user_service.entity.goal.GoalStatus.ACTIVE;
+import static school.faang.user_service.entity.goal.GoalStatus.COMPLETED;
+
+@Component
+@RequiredArgsConstructor
+public class GoalValidator {
+    @Value("${app.goal.max-active-per-user}")
+    private Integer maxExistedActiveGoals;
+
+    private final UserRepository userRepository;
+    private final GoalRepository goalRepository;
+    private final SkillRepository skillRepository;
```
Нет. 
Валидатор не должен ни в какую базу ходить и что-то из неё тащить для проверки, эта вся логика должна быть в сервисах.

Валидация - проверка полей на смысловую корректность относительно доменной области.

amilosh replied:
Перенес в сервисы.

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+
+    @PostMapping
+    @ResponseStatus(CREATED)
+    public void createGoal(@RequestBody @Validated(Create.class) GoalDto goalDto) {
```
@Validated(Create.class) - что это?

20cognosce commented on src/main/java/school/faang/user_service/dto/validation/group/Create.java:
```diff
@@ -0,0 +1,4 @@
+package school.faang.user_service.dto.validation.group;
+
+public interface Create {
```
?

20cognosce commented on src/main/java/school/faang/user_service/dto/validation/group/Update.java:
```diff
@@ -0,0 +1,4 @@
+package school.faang.user_service.dto.validation.group;
+
+public interface Update {
```
?

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @PutMapping
+    public void updateGoal(@RequestBody @Validated(Update.class) GoalDto goalDto) {
```
Что это за маркерные интерфейсы для круда, зачем они?

20cognosce replied:
@PatchMapping

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+    @PutMapping
+    public void updateGoal(@RequestBody @Validated(Update.class) GoalDto goalDto) {
+        goalValidator.validateUpdating(goalDto);
+        var goal = goalMapper.toEntity(goalDto);
```
Прошу максимально редко юзать var

amilosh replied:
убрал

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @GetMapping("/sub-goals/{parentGoalId}")
+    public List<GoalDto> findSubGoalsByParentGoalId(@PathVariable Long parentGoalId, GoalFilterDto filter) {
```
Сейчас у тебя 3 параметра фильтрации, а потом будет 23
И это всё будет в урле?
Давай в тело класть и через Post-запрос делать

В урле я за то чтобы передавать только общеизвестные параметры страницы - number, asc/desc

Всё что относится к доменной области лучше прятать от юзера

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @GetMapping("/user/{userId}")
+    public List<GoalDto> findGoalsByUser(@PathVariable long userId, GoalFilterDto filter) {
```
Аналогично

20cognosce commented on src/main/java/school/faang/user_service/mapping/GoalMapper.java:
```diff
@@ -0,0 +1,67 @@
+    unmappedTargetPolicy = IGNORE)
+public abstract class GoalMapper {
+
+    @Autowired
+    protected GoalRepository goalRepository;
+
+    @Autowired
+    protected SkillRepository skillRepository;
```
1. @Autowired - может никак не предупредить если бин окажется null, лучше конструктор для надежности
2. Почему твой маппер общается с репозиториями? Это разный логический уровень приложения.
Маппер DTOшек это уровень представления
Репозитории это уровень данных

protected Goal mapGoalIdToGoal(Long id) - такого тут вообще не должно быть

Что может смапить - мапь, что не можешь, например id - передавай их в сервис

amilosh replied:
Сделал.

20cognosce commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -11,20 +12,34 @@
             """)
     Goal create(String title, String description, Long parent);
 
+    @Modifying
+    @Query(nativeQuery = true, value = """
+            UPDATE goal 
+            SET title = :title, description = :description, parent_goal_id = :parent, status = :status, updated_at = NOW()
+            WHERE id = :goalId
+            """)
+    int update(Long goalId, String title, String description, Long parent, int status);
```
Почему ты все CRUDы переписал на нативные запросы??
Используй готовые методы JpaRepository, свои собственные запросы надо писать когда есть тонкая логика какая-то, а такой апдейта - это громоздко и избыточно

Другие методы тоже упростить и использовать готовое

amilosh replied:
Сделал.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public void createGoal(Goal goal, Long userId) {
+        Goal createdGoal = goalRepository.create(
+            goal.getTitle(), goal.getDescription(), goal.getParent() == null ? null : goal.getParent().getId());
```
.create(goal);

переделать

amilosh replied:
сделал через save(goal)

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public void updateGoal(Goal goal) {
+        goalRepository.update(goal.getId(), goal.getTitle(), goal.getDescription(),
```
update(goal);

amilosh replied:
сделал через findById
потом save

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+    @Override
+    public void updateGoal(Goal goal) {
+        goalRepository.update(goal.getId(), goal.getTitle(), goal.getDescription(),
+            goal.getParent() == null ? null : goal.getParent().getId(), goal.getStatus().ordinal());
+
+        updateSkills(goal);
+
+        if (isMakeGoalCompleted(goal)) {
+            assignGoalSkillsToUsers(goal);
+        }
```
Что тут происходит?
Ты обновил цель, и потом что-то происходит, совсем не очевидное что

+Транзакция нужна, если это связанные действия

amilosh replied:
переделал

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public void deleteGoal(Long goalId) {
+        List<Long> goalIds = new ArrayList<>();
+        goalRepository.findByParent(goalId)
+            .toList()
+            .forEach(goal -> goalIds.add(goal.getId()));
+        goalIds.add(goalId);
+
+        for (Long id : goalIds) {
+            goalRepository.unassignGoalFromUsers(id);
+            skillRepository.findSkillsByGoalId(id)
+                .forEach(skill -> skillRepository.unassignSkillFromUsers(skill.getId()));
+            skillRepository.deleteSkillsByGoalId(id);
+
+            goalRepository.delete(id);
+        }
```
Аналогично, что всё это?
Проще должно быть и понятнее

Если связанные действия - транзакция 
@Transactional

amilosh replied:
переделал

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+        return filterGoalsByPredicate(goalsByUser, predicate);
+    }
+
+    private Goal findById(Long goalId) {
+        return goalRepository.findById(goalId)
+            .orElseThrow(() -> new ResourceNotFoundException("Goal", goalId));
+    }
```
Должен быть публичный метод поиска цели

amilosh replied:
сделал

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,131 @@
+            .orElseThrow(() -> new ResourceNotFoundException("Goal", goalId));
+    }
+
+    private void updateSkills(Goal goal) {
+        skillRepository.deleteSkillsByGoalId(goal.getId());
+        assignSkillsToGoal(goal, goal.getId());
+    }
+
+    private void assignSkillsToGoal(Goal goal, Long goalId) {
+        List<Skill> skillsToAchieve = goal.getSkillsToAchieve();
+        if (isNotEmpty(skillsToAchieve)) {
+            List<Long> skillIds = skillsToAchieve.stream()
+                .map(Skill::getId)
+                .toList();
+            skillIds.forEach(skillId -> skillRepository.addSkillToGoal(skillId, goalId));
+        }
+    }
+
+    private void assignGoalSkillsToUsers(Goal goal) {
+        List<Skill> skills = skillRepository.findSkillsByGoalId(goal.getId());
+        Goal foundGoal = goalRepository.findById(goal.getId()).get();
+        foundGoal.getUsers().forEach(user -> {
+            skills.forEach(skill -> skillRepository.assignSkillToUser(skill.getId(), user.getId()));
+        });
+    }
+
+    private boolean isMakeGoalCompleted(Goal goal) {
+        Goal goalFromDb = findById(goal.getId());
+        return goal.getStatus() == COMPLETED && goalFromDb.getStatus() == ACTIVE;
+    }
```
По максимуму это вынести в публичные методы чтобы не было так, что у тебя 3 публичных метода по одной строчке

и 23 приватных метода которые переворачивают всё с вверх головой, что ничего не разобрать, что происходит

amilosh replied:
вынес

20cognosce commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,243 @@
+
+        goalService.updateGoal(goal);
+
+        verify(goalRepository).findById(eq(GOAL_ID));
+        verify(goalRepository).update(eq(GOAL_ID), eq(GOAL_TITLE), eq(GOAL_DESCRIPTION),
+            eq(null), eq(goal.getStatus().ordinal()));
+        verify(skillRepository).deleteSkillsByGoalId(eq(GOAL_ID));
+        verify(skillRepository).addSkillToGoal(eq(SKILL_ID), eq(GOAL_ID));
+        verify(skillRepository, times(0)).findSkillsByGoalId(anyLong());
+        verify(skillRepository, times(0)).assignSkillToUser(anyLong(), anyLong());
```
Что это за проверки?
Ты должен тестировать метод как черный ящик - только сравнение выходного значения с ожидаемым результатом, да, на crud'ах тут нет почти никакой логики, но тем не менее - times(0) - это то, что мне не нравится.

Зачем проверять, что что-то не вызывалось? Это нелогично

amilosh replied:
убрал

amilosh-softeq committed:
BJS2-27008: make some updates after mentor review

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,63 @@
+    @ResponseStatus(CREATED)
+    public GoalResponseDto createGoal(@RequestBody @Valid CreateGoalDto dto) {
+        Goal goal = goalMapper.toEntity(dto);
+        Goal createdGoal = goalService.createGoal(goal, dto.getUserId(), dto.getParentGoalId(), dto.getSkillIds());
```
Лучше объявлять переменные

amilosh replied:
Вынес.

20cognosce commented on src/main/java/school/faang/user_service/entity/goal/Goal.java:
```diff
@@ -13,6 +13,8 @@
 import java.time.LocalDateTime;
 import java.util.List;
 
+import static school.faang.user_service.entity.goal.GoalStatus.ACTIVE;
```
?

amilosh replied:
Экспериментировал.
Не доглядел убрать.

20cognosce commented on src/main/java/school/faang/user_service/exception/constants/ErrorMessage.java:
```diff
@@ -0,0 +1,8 @@
+package school.faang.user_service.exception.constants;
+
+public final class ErrorMessage {
+    private ErrorMessage() {
+    }
```
@UtilityClass

amilosh replied:
Заюзал.

20cognosce commented on src/main/java/school/faang/user_service/exception/handler/RestExceptionHandler.java:
```diff
@@ -0,0 +1,34 @@
+
+    private ResponseEntity<?> handleExceptionInternal(Exception ex, HttpStatus status) {
+        log.error(status.name(), ex);
+        var errorResponse = ErrorResponse.builder()
```
var - лучше избегать и писать полностью класс

amilosh replied:
Убрал.
Больше не буду.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+            throw new BadRequestException("User %s can have maximum %s goals", userId, maxExistedActiveGoals);
+        }
+
+        if (parentGoalId != null) {
+            goalRepository.findById(parentGoalId)
+                .orElseThrow(() -> new ResourceNotFoundException("Parent Goal", parentGoalId));
+        }
```
Почему у тебя в этом классе есть еще метод findById, не находишь это дублированием кода?

amilosh replied:
Да, есть такой метод. Но ошибка "Goal not found". Я хочу бросить. что именно "Parent Goal not found".
Создал ещё один метод со вторым String параметром.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+            orElseThrow(() -> new ResourceNotFoundException("User", userId));
+        goal.setUsers(List.of(user));
+
+        Goal parentGoal = parentGoalId != null ? findById(parentGoalId) : null;
```
Ты это уже проверял

amilosh replied:
Вынес засечивание parentGoal и его проверку в одно место.
Стало красивше.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+        Goal parentGoal = parentGoalId != null ? findById(parentGoalId) : null;
+        goal.setParent(parentGoal);
+
+        List<Skill> skills = skillIds != null ? skillRepository.findByIds(skillIds) : null;
```
Почему в валидации нельзя все null проверить и очистить метод этот этого?

amilosh replied:
Для skillIds не надо на null валидировать. Там только существование килов в базе проверка. skillIds могут быть как null так какие-то значения.
И если skillIds - null, то и в Gaol set skills тоже null подставлять.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+        }
+        boolean isMakeGoalCompleted = goal.getStatus() == COMPLETED && existedGoal.getStatus() == ACTIVE;
+
+        goalMapper.update(goal, existedGoal);
```
Не понял. Почему маппер управляет обновлением сущностей?
Делай всё тут руками

amilosh replied:
Поменял на сеттеры.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Override
+    public Goal updateGoal(Goal goal, List<Long> skillIds) {
```
Можно порефакторить эти методы? сделать более компактными и структурированными

20cognosce replied:
Транзакция нужна -  @Transactional

amilosh replied:
Не понимаю как.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Override
+    public void deleteGoal(Long goalId) {
```
Транзакция

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+    public void deleteGoal(Long goalId) {
+        goalRepository.findByParent(goalId)
+            .toList()
+            .forEach(goal -> goalRepository.deleteById(goal.getId()));
```
Интересно, а если у детей этой goalId есть свои дети?
Кажется тут нужна неограниченная вложенность, пока у родителя есть дети

amilosh replied:
Я это на потом оставил.
Мне надо день потратить, чтобы понять как по рекурсии все удалить.

amilosh replied:
А всё, нашел. У нас уже есть готовые запросы с рекурсией.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+    @Transactional(readOnly = true)
+    public List<Goal> findSubGoalsByParentGoalId(Long parentGoalId, GoalFilterDto filterDto) {
+        Predicate<Goal> predicate = combinePredicateFromFilter(filterDto);
+        Stream<Goal> subGoalsByParentGoal = goalRepository.findByParent(parentGoalId);
```
Почему репозиторий возвращает стрим? Это зачем

20cognosce replied:
Лучше на список переделать и только тут брать стрим

amilosh replied:
Добавил методы в репозиторий, чтобы вернуть List.

20cognosce commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,271 @@
+            goalService.createGoal(goal, USER_ID, null, List.of(SKILL_ID))
+        );
+
+        assertEquals("Skills from request are not presented in DB", exception.getMessage());
```
Сообщение можно не проверять мне кажется, просто класс исключения

amilosh replied:
Убрал сообщения про Skills.
Где фигурирует Goal id - оставил.

20cognosce commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,271 @@
+
+        Goal result = goalService.createGoal(goal, USER_ID, PARENT_GOAL_ID, List.of(SKILL_ID));
+
+        assertEquals(createdGoal.getId(), result.getId());
```
А почему только id?
AssertThat() recursive comparison
по всем полям попробуй

amilosh replied:
Прикольная штука. Буду знать.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+import static school.faang.user_service.entity.goal.GoalStatus.COMPLETED;
+
+@Service
+@Transactional
```
А, ты на весь класс навесил
Это не хорошо - любой публичный метод в транзакцию пихать автоматически, ну такое

amilosh replied:
Перенес только на методы, которые вызываются в конроллере.

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+    private final SkillRepository skillRepository;
+    private final UserRepository userRepository;
+    private final List<GoalFilter> goalFilters;
+    private final GoalMapper goalMapper;
```
Вообще не должно быть зависимости на маппер у сервиса
Это разные логические слои

amilosh replied:
Убрал.

amilosh-softeq committed:
BJS2-27008: make some changes after mentor review

20cognosce commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -43,6 +50,17 @@ WITH RECURSIVE subtasks AS (
     Stream<Goal> findByParent(long goalId);
 
+    @Query(nativeQuery = true, value = """
+            WITH RECURSIVE subtasks AS (
+            SELECT * FROM goal WHERE id = :goalId
+            UNION
+            SELECT g.* FROM goal g
+            JOIN subtasks st ON st.id = g.parent_goal_id
+            )
+            SELECT * FROM subtasks WHERE id != :goalId
+            """)
```
Выглядит жестко
Надеюсь работает как задумано

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        User user = userRepository.findById(userId).
+            orElseThrow(() -> new ResourceNotFoundException("User", userId));
```
. со следующей строки
